core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -535,8 +531,13 @@ public void handleException(TransportException e) {
                                         }
                                         finally {
                                             latch.countDown();
+                                            closeConnection();
                                         }
                                     }
+
+                                    void closeConnection() {
+                                        IOUtils.closeWhileHandlingException(connectionToClose);
+                                    }
```

Comment:
I wrote a test which would have failed before the fix. That would be great if you can have a look.